File: 6. ZigZag Conversion/convert.ts
function convert(s: string, numRows: number): string {
  if (numRows === 1) return s;
  let rows: string[] = new Array(numRows).fill('');
  let row: number = 0;
  let up: boolean = true;

  for (let char of s) {
    rows[row] += char;
    row = up ? row + 1 : row - 1;
    if (row === 0) up = true;
    if (row === numRows - 1) up = false;
  }

  return rows.join('');
};

console.log(convert('PAYPALISHIRING', 3));
console.log(convert('PAYPALISHIRING', 1));
console.log(convert('PAYPALISHIRING', 4));

File: 6. ZigZag Conversion/convert.rb
def convert(s, num_rows)
  return s if num_rows == 1

  rows = [''] * num_rows
  row = 0
  up = true

  s.each_char do |char|
    rows[row] += char
    row = up ? row + 1 : row - 1
    up = false if  row == num_rows - 1
    up = true if row == 0
  end

  rows.join ''  
end

pp convert 'PAYPALISHIRING', 3 # PAHNAPLSIIGYIR
pp convert 'PAYPALISHIRING', 4 # PINALSIGYAHRPI

